### DISCUSSION TITLE
Accessing sensors etc

### DISCUSSION THREAD

**Discussion by perceptualrobots** [cyberbotics/wrestling#37] [Q&A] [answered]
In the Nao picture in poster.pdf I see the sensors and actuators.

Perhaps, I am mssing something, but how do I use them? That is, how do I read from the sensors and write values to the actuators.

Is there documentation that explains this?I looked here, but didn't see what I was looking for [https://cyberbotics.com/doc/guide/nao](https://cyberbotics.com/doc/guide/nao).

Also, are we able, during the loop to read performance evaluation values, such as coverage, position, height and participant scores?

**Comment by zicklag** [comment, marked as answer]
> Also, are we able, during the loop to read performance evaluation values, such as coverage, position, height and participant scores?

Somebody may correct me if I'm wrong, but I believe the robot controller is not able to read those metrics. Only the sensors actually available to the NAO robot are available to your robot controller.

> Perhaps, I am mssing something, but how do I use them? That is, how do I read from the sensors and write values to the actuators.

You do that using the webots Robot API. Most of the examples are in C or in Python.

You can find a Python example here:

https://github.com/cyberbotics/wrestling-bob/blob/14df242412ff75d50c664560afc967d620f1a43c/controllers/participant/participant.py#L26-L41

This shows getting the motors for the left and right shoulders, and setting their position. 

I think you can get to the full API docs at: https://cyberbotics.com/doc/reference/robot.

**Reply by Jean-Eudes-le-retour** [reply]
> Somebody may correct me if I'm wrong, but I believe the robot controller is not able to read those metrics. Only the sensors actually available to the NAO robot are available to your robot controller.

This is correct.

> how do I read from the sensors and write values to the actuators.

Here is a guide on how to get started programming a controller, you should be able to find all the information you need:
https://cyberbotics.com/doc/guide/controller-programming

**Reply by perceptualrobots** [reply]
> Somebody may correct me if I'm wrong, but I believe the robot controller is not able to read those metrics. Only the sensors actually available to the NAO robot are available to your robot controller.

I am running a genetic algorithm which relies on an evaluation of each run. Surely there must be some way of knowing how well a controller has done, even if from outside the scope of the controller? 

> You do that using the webots Robot API. Most of the examples are in C or in Python.
> You can find a Python example here:
> https://github.com/cyberbotics/wrestling-bob/blob/14df242412ff75d50c664560afc967d620f1a43c/controllers/participant/participant.py#L26-L41
> This shows getting the motors for the left and right shoulders, and setting their position.
> I think you can get to the full API docs at: https://cyberbotics.com/doc/reference/robot.

The example shows an actuator called RShoulderPitch. How do I know the robot has a device with that name?

The API shows a method called getDevice, but how do I know what sensors/actuators the robot has and what they are called?

**Reply by Jean-Eudes-le-retour** [reply]
The sensors and actuators are defined inside the robot's model. You can find more information on which sensors and actuators are present in the robot on these links:
- https://cyberbotics.com/doc/guide/nao
- [Dedicated proto page](https://webots.cloud/run?version=R2023b&url=https%3A%2F%2Fgithub.com%2Fcyberbotics%2Fwebots%2Fblob%2Fdevelop%2Fprojects%2Frobots%2Fsoftbank%2Fnao%2Fprotos%2FNao.proto&type=undefined)

**Reply by omichel** [reply]
To know how well your controller performs and to optimize it using a generic algorithm or anything else, you should run the simulation on your local computer and modify the referee supervisor to send the performance evaluation and other information back to the controller, so that you can close the evolutionary loop.
